### PR TITLE
Validate that snapshot repository exists for ILM policies during GenerateSnapshotNameStep

### LINKED ISSUE #72957: ILM searchable snapshot repo change does not affect existing indices
**Elasticsearch version** : 7.12.1, ECE - 2.9.1

**Description of the problem including expected versus actual behavior**: 

ILM repo change for searchable snapshot action does not affect an existing index

**Steps to reproduce**:

 1. Create an ILM policy with frozen phase, having a searchable snapshot, set a wrong repo for snapshots
 2. ILM will fail and complain of incorrect repo and it will wait for repo to be corrected and will keep retrying, this can be checked via `<index>/_ilm/explain`
 3. Update ILM policy and set the correct repo name for the searchable snapshot action in the frozen phase
 4. It should auto-retry, if not Hit `POST <index>/_ilm/retry`
 5. Even though the repo is corrected in ILM, it will still keep erroring out on incorrect/old repo

**Provide logs (if relevant)**:

```
{
  "indices" : {
    "metricbeat-7.6.0-2021-05-04-001068" : {
      "index" : "metricbeat-7.6.0-2021-05-04-001068",
      "managed" : true,
      "policy" : "metrics-ss",
      "lifecycle_date_millis" : 1620742701627,
      "age" : "5.33h",
      "phase" : "frozen",
      "phase_time_millis" : 1620761901282,
      "action" : "searchable_snapshot",
      "action_time_millis" : 1620743117857,
      "step" : "ERROR",
      "step_time_millis" : 1620761901428,
      "failed_step" : "cleanup-snapshot",
      "is_auto_retryable_error" : true,
      "failed_step_retry_count" : 7,
      "repository_name" : "searchable-snapshots-712",
      "snapshot_name" : "2021.05.11-metricbeat-7.6.0-2021-05-04-001068-metrics-ss-enfipo8vtfc0761m1gjccq",
      "step_info" : {
        "type" : "illegal_state_exception",
        "reason" : "repository [searchable-snapshots-712] is missing. [metrics-ss] policy for index [metricbeat-7.6.0-2021-05-04-001068] cannot continue until the repository is created",
        "caused_by" : {
          "type" : "repository_missing_exception",
          "reason" : "[searchable-snapshots-712] missing",
          "stack_trace" : """RepositoryMissingException[[searchable-snapshots-712] missing]
	at org.elasticsearch.repositories.RepositoriesService.repository(RepositoriesService.java:504)
	at org.elasticsearch.snapshots.SnapshotsService.deleteSnapshots(SnapshotsService.java:1990)
	at org.elasticsearch.action.admin.cluster.snapshots.delete.TransportDeleteSnapshotAction.masterOperation(TransportDeleteSnapshotAction.java:49)
	at org.elasticsearch.action.admin.cluster.snapshots.delete.TransportDeleteSnapshotAction.masterOperation(TransportDeleteSnapshotAction.java:28)
	at org.elasticsearch.action.support.master.TransportMasterNodeAction.masterOperation(TransportMasterNodeAction.java:88)
	at org.elasticsearch.action.support.master.TransportMasterNodeAction$AsyncSingleAction.lambda$doStart$3(TransportMasterNodeAction.java:155)
	at org.elasticsearch.action.ActionRunnable$2.doRun(ActionRunnable.java:62)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	at org.elasticsearch.common.util.concurrent.EsExecutors$DirectExecutorService.execute(EsExecutors.java:215)
	at org.elasticsearch.action.support.master.TransportMasterNodeAction$AsyncSingleAction.doStart(TransportMasterNodeAction.java:155)
	at org.elasticsearch.action.support.master.TransportMasterNodeAction.doExecute(TransportMasterNodeAction.java:104)
	at org.elasticsearch.action.support.master.TransportMasterNodeAction.doExecute(TransportMasterNodeAction.java:46)
	at org.elasticsearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:173)
	at org.elasticsearch.action.support.ActionFilter$Simple.apply(ActionFilter.java:42)
	at org.elasticsearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:171)
	at org.elasticsearch.xpack.security.action.filter.SecurityActionFilter.lambda$applyInternal$3(SecurityActionFilter.java:161)
	at org.elasticsearch.action.ActionListener$3.onResponse(ActionListener.java:167)
	at org.elasticsearch.xpack.security.authz.AuthorizationService.lambda$authorizeAction$4(AuthorizationService.java:280)
	at org.elasticsearch.xpack.security.authz.AuthorizationService$AuthorizationResultListener.onResponse(AuthorizationService.java:684)
	at org.elasticsearch.xpack.security.authz.AuthorizationService$AuthorizationResultListener.onResponse(AuthorizationService.java:659)
	at org.elasticsearch.action.support.ContextPreservingActionListener.onResponse(ContextPreservingActionListener.java:32)
	at org.elasticsearch.xpack.security.authz.RBACEngine.authorizeClusterAction(RBACEngine.java:154)
	at org.elasticsearch.xpack.security.authz.AuthorizationService.authorizeAction(AuthorizationService.java:282)
	at org.elasticsearch.xpack.security.authz.AuthorizationService.maybeAuthorizeRunAs(AuthorizationService.java:265)
	at org.elasticsearch.xpack.security.authz.AuthorizationService.lambda$authorize$1(AuthorizationService.java:229)
	at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:117)
	at org.elasticsearch.action.support.ContextPreservingActionListener.onResponse(ContextPreservingActionListener.java:32)
	at org.elasticsearch.xpack.security.authz.RBACEngine.lambda$resolveAuthorizationInfo$1(RBACEngine.java:127)
	at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:117)
	at org.elasticsearch.xpack.security.authz.store.CompositeRolesStore.getRoles(CompositeRolesStore.java:276)
	at org.elasticsearch.xpack.security.authz.RBACEngine.getRoles(RBACEngine.java:133)
	at org.elasticsearch.xpack.security.authz.RBACEngine.resolveAuthorizationInfo(RBACEngine.java:121)
	at org.elasticsearch.xpack.security.authz.AuthorizationService.authorize(AuthorizationService.java:231)
	at org.elasticsearch.xpack.security.action.filter.SecurityActionFilter.authorizeRequest(SecurityActionFilter.java:181)
	at org.elasticsearch.xpack.security.action.filter.SecurityActionFilter.lambda$applyInternal$4(SecurityActionFilter.java:159)
	at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:117)
	at org.elasticsearch.xpack.security.authc.AuthenticationService$Authenticator.lambda$authenticateAsync$2(AuthenticationService.java:330)
	at org.elasticsearch.xpack.security.authc.AuthenticationService$Authenticator.lambda$lookForExistingAuthentication$6(AuthenticationService.java:391)
	at org.elasticsearch.xpack.security.authc.AuthenticationService$Authenticator.lookForExistingAuthentication(AuthenticationService.java:402)
	at org.elasticsearch.xpack.security.authc.AuthenticationService$Authenticator.authenticateAsync(AuthenticationService.java:327)
	at org.elasticsearch.xpack.security.authc.AuthenticationService$Authenticator.access$000(AuthenticationService.java:268)
	at org.elasticsearch.xpack.security.authc.AuthenticationService.authenticate(AuthenticationService.java:161)
	at org.elasticsearch.xpack.security.action.filter.SecurityActionFilter.applyInternal(SecurityActionFilter.java:154)
	at org.elasticsearch.xpack.security.action.filter.SecurityActionFilter.apply(SecurityActionFilter.java:106)
	at org.elasticsearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:171)
	at org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:149)
	at org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:77)
	at org.elasticsearch.client.node.NodeClient.executeLocally(NodeClient.java:86)
	at org.elasticsearch.client.node.NodeClient.doExecute(NodeClient.java:66)
	at org.elasticsearch.client.support.AbstractClient.execute(AbstractClient.java:402)
	at org.elasticsearch.xpack.core.ClientHelper.executeWithHeadersAsync(ClientHelper.java:196)
	at org.elasticsearch.xpack.ilm.LifecyclePolicySecurityClient.doExecute(LifecyclePolicySecurityClient.java:52)
	at org.elasticsearch.client.support.AbstractClient.execute(AbstractClient.java:402)
	at org.elasticsearch.client.support.AbstractClient$ClusterAdmin.execute(AbstractClient.java:695)
	at org.elasticsearch.client.support.AbstractClient$ClusterAdmin.deleteSnapshot(AbstractClient.java:997)
	at org.elasticsearch.xpack.core.ilm.CleanupSnapshotStep.performDuringNoSnapshot(CleanupSnapshotStep.java:54)
	at org.elasticsearch.xpack.core.ilm.AsyncRetryDuringSnapshotActionStep.performAction(AsyncRetryDuringSnapshotActionStep.java:42)
	at org.elasticsearch.xpack.ilm.IndexLifecycleRunner.maybeRunAsyncAction(IndexLifecycleRunner.java:290)
	at org.elasticsearch.xpack.ilm.IndexLifecycleRunner$2.clusterStateProcessed(IndexLifecycleRunner.java:246)
	at org.elasticsearch.cluster.service.MasterService$SafeClusterStateTaskListener.clusterStateProcessed(MasterService.java:523)
	at org.elasticsearch.cluster.service.MasterService$TaskOutputs.lambda$processedDifferentClusterState$1(MasterService.java:410)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.elasticsearch.cluster.service.MasterService$TaskOutputs.processedDifferentClusterState(MasterService.java:410)
	at org.elasticsearch.cluster.service.MasterService.onPublicationSuccess(MasterService.java:270)
	at org.elasticsearch.cluster.service.MasterService.publish(MasterService.java:262)
	at org.elasticsearch.cluster.service.MasterService.runTasks(MasterService.java:239)
	at org.elasticsearch.cluster.service.MasterService.access$000(MasterService.java:62)
	at org.elasticsearch.cluster.service.MasterService$Batcher.run(MasterService.java:140)
	at org.elasticsearch.cluster.service.TaskBatcher.runIfNotProcessed(TaskBatcher.java:139)
	at org.elasticsearch.cluster.service.TaskBatcher$BatchedTask.run(TaskBatcher.java:177)
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:673)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:241)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:204)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at java.base/java.lang.Thread.run(Thread.java:831)
"""
        },
        "stack_trace" : """java.lang.IllegalStateException: repository [searchable-snapshots-712] is missing. [metrics-ss] policy for index [metricbeat-7.6.0-2021-05-04-001068] cannot continue until the repository is created
	at org.elasticsearch.xpack.core.ilm.CleanupSnapshotStep$1.onFailure(CleanupSnapshotStep.java:74)
	at org.elasticsearch.action.support.ContextPreservingActionListener.onFailure(ContextPreservingActionListener.java:39)
	at org.elasticsearch.action.support.TransportAction$1.onFailure(TransportAction.java:92)
	at org.elasticsearch.action.support.ContextPreservingActionListener.onFailure(ContextPreservingActionListener.java:39)
	at org.elasticsearch.action.ActionListener$3.onFailure(ActionListener.java:172)
	at org.elasticsearch.action.support.master.TransportMasterNodeAction$AsyncSingleAction.lambda$doStart$2(TransportMasterNodeAction.java:151)
	at org.elasticsearch.action.ActionListener$2.onFailure(ActionListener.java:148)
	at org.elasticsearch.action.ActionRunnable.onFailure(ActionRunnable.java:77)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:28)
	at org.elasticsearch.common.util.concurrent.EsExecutors$DirectExecutorService.execute(EsExecutors.java:215)
	at org.elasticsearch.action.support.master.TransportMasterNodeAction$AsyncSingleAction.doStart(TransportMasterNodeAction.java:155)
	at org.elasticsearch.action.support.master.TransportMasterNodeAction.doExecute(TransportMasterNodeAction.java:104)
	at org.elasticsearch.action.support.master.TransportMasterNodeAction.doExecute(TransportMasterNodeAction.java:46)
	at org.elasticsearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:173)
	at org.elasticsearch.action.support.ActionFilter$Simple.apply(ActionFilter.java:42)
	at org.elasticsearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:171)
	at org.elasticsearch.xpack.security.action.filter.SecurityActionFilter.lambda$applyInternal$3(SecurityActionFilter.java:161)
	at org.elasticsearch.action.ActionListener$3.onResponse(ActionListener.java:167)
	at org.elasticsearch.xpack.security.authz.AuthorizationService.lambda$authorizeAction$4(AuthorizationService.java:280)
	at org.elasticsearch.xpack.security.authz.AuthorizationService$AuthorizationResultListener.onResponse(AuthorizationService.java:684)
	at org.elasticsearch.xpack.security.authz.AuthorizationService$AuthorizationResultListener.onResponse(AuthorizationService.java:659)
	at org.elasticsearch.action.support.ContextPreservingActionListener.onResponse(ContextPreservingActionListener.java:32)
	at org.elasticsearch.xpack.security.authz.RBACEngine.authorizeClusterAction(RBACEngine.java:154)
	at org.elasticsearch.xpack.security.authz.AuthorizationService.authorizeAction(AuthorizationService.java:282)
	at org.elasticsearch.xpack.security.authz.AuthorizationService.maybeAuthorizeRunAs(AuthorizationService.java:265)
	at org.elasticsearch.xpack.security.authz.AuthorizationService.lambda$authorize$1(AuthorizationService.java:229)
	at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:117)
	at org.elasticsearch.action.support.ContextPreservingActionListener.onResponse(ContextPreservingActionListener.java:32)
	at org.elasticsearch.xpack.security.authz.RBACEngine.lambda$resolveAuthorizationInfo$1(RBACEngine.java:127)
	at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:117)
	at org.elasticsearch.xpack.security.authz.store.CompositeRolesStore.getRoles(CompositeRolesStore.java:276)
	at org.elasticsearch.xpack.security.authz.RBACEngine.getRoles(RBACEngine.java:133)
	at org.elasticsearch.xpack.security.authz.RBACEngine.resolveAuthorizationInfo(RBACEngine.java:121)
	at org.elasticsearch.xpack.security.authz.AuthorizationService.authorize(AuthorizationService.java:231)
	at org.elasticsearch.xpack.security.action.filter.SecurityActionFilter.authorizeRequest(SecurityActionFilter.java:181)
	at org.elasticsearch.xpack.security.action.filter.SecurityActionFilter.lambda$applyInternal$4(SecurityActionFilter.java:159)
	at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:117)
	at org.elasticsearch.xpack.security.authc.AuthenticationService$Authenticator.lambda$authenticateAsync$2(AuthenticationService.java:330)
	at org.elasticsearch.xpack.security.authc.AuthenticationService$Authenticator.lambda$lookForExistingAuthentication$6(AuthenticationService.java:391)
	at org.elasticsearch.xpack.security.authc.AuthenticationService$Authenticator.lookForExistingAuthentication(AuthenticationService.java:402)
	at org.elasticsearch.xpack.security.authc.AuthenticationService$Authenticator.authenticateAsync(AuthenticationService.java:327)
	at org.elasticsearch.xpack.security.authc.AuthenticationService$Authenticator.access$000(AuthenticationService.java:268)
	at org.elasticsearch.xpack.security.authc.AuthenticationService.authenticate(AuthenticationService.java:161)
	at org.elasticsearch.xpack.security.action.filter.SecurityActionFilter.applyInternal(SecurityActionFilter.java:154)
	at org.elasticsearch.xpack.security.action.filter.SecurityActionFilter.apply(SecurityActionFilter.java:106)
	at org.elasticsearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:171)
	at org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:149)
	at org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:77)
	at org.elasticsearch.client.node.NodeClient.executeLocally(NodeClient.java:86)
	at org.elasticsearch.client.node.NodeClient.doExecute(NodeClient.java:66)
	at org.elasticsearch.client.support.AbstractClient.execute(AbstractClient.java:402)
	at org.elasticsearch.xpack.core.ClientHelper.executeWithHeadersAsync(ClientHelper.java:196)
	at org.elasticsearch.xpack.ilm.LifecyclePolicySecurityClient.doExecute(LifecyclePolicySecurityClient.java:52)
	at org.elasticsearch.client.support.AbstractClient.execute(AbstractClient.java:402)
	at org.elasticsearch.client.support.AbstractClient$ClusterAdmin.execute(AbstractClient.java:695)
	at org.elasticsearch.client.support.AbstractClient$ClusterAdmin.deleteSnapshot(AbstractClient.java:997)
	at org.elasticsearch.xpack.core.ilm.CleanupSnapshotStep.performDuringNoSnapshot(CleanupSnapshotStep.java:54)
	at org.elasticsearch.xpack.core.ilm.AsyncRetryDuringSnapshotActionStep.performAction(AsyncRetryDuringSnapshotActionStep.java:42)
	at org.elasticsearch.xpack.ilm.IndexLifecycleRunner.maybeRunAsyncAction(IndexLifecycleRunner.java:290)
	at org.elasticsearch.xpack.ilm.IndexLifecycleRunner$2.clusterStateProcessed(IndexLifecycleRunner.java:246)
	at org.elasticsearch.cluster.service.MasterService$SafeClusterStateTaskListener.clusterStateProcessed(MasterService.java:523)
	at org.elasticsearch.cluster.service.MasterService$TaskOutputs.lambda$processedDifferentClusterState$1(MasterService.java:410)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.elasticsearch.cluster.service.MasterService$TaskOutputs.processedDifferentClusterState(MasterService.java:410)
	at org.elasticsearch.cluster.service.MasterService.onPublicationSuccess(MasterService.java:270)
	at org.elasticsearch.cluster.service.MasterService.publish(MasterService.java:262)
	at org.elasticsearch.cluster.service.MasterService.runTasks(MasterService.java:239)
	at org.elasticsearch.cluster.service.MasterService.access$000(MasterService.java:62)
	at org.elasticsearch.cluster.service.MasterService$Batcher.run(MasterService.java:140)
	at org.elasticsearch.cluster.service.TaskBatcher.runIfNotProcessed(TaskBatcher.java:139)
	at org.elasticsearch.cluster.service.TaskBatcher$BatchedTask.run(TaskBatcher.java:177)
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:673)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:241)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:204)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at java.base/java.lang.Thread.run(Thread.java:831)
Caused by: RepositoryMissingException[[searchable-snapshots-712] missing]
	at org.elasticsearch.repositories.RepositoriesService.repository(RepositoriesService.java:504)
	at org.elasticsearch.snapshots.SnapshotsService.deleteSnapshots(SnapshotsService.java:1990)
	at org.elasticsearch.action.admin.cluster.snapshots.delete.TransportDeleteSnapshotAction.masterOperation(TransportDeleteSnapshotAction.java:49)
	at org.elasticsearch.action.admin.cluster.snapshots.delete.TransportDeleteSnapshotAction.masterOperation(TransportDeleteSnapshotAction.java:28)
	at org.elasticsearch.action.support.master.TransportMasterNodeAction.masterOperation(TransportMasterNodeAction.java:88)
	at org.elasticsearch.action.support.master.TransportMasterNodeAction$AsyncSingleAction.lambda$doStart$3(TransportMasterNodeAction.java:155)
	at org.elasticsearch.action.ActionRunnable$2.doRun(ActionRunnable.java:62)
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
	... 68 more
"""
      },
      "phase_execution" : {
        "policy" : "metrics-ss",
        "phase_definition" : {
          "min_age" : "6m",
          "actions" : {
            "searchable_snapshot" : {
              "snapshot_repository" : "searchable-snapshots",
              "force_merge_index" : true
            }
          }
        },
        "version" : 4,
        "modified_date_in_millis" : 1620752356149
      }
    }
  }
}

```

### PR DESCRIPTION
**Pull request by joegallo**
Fixes #72957

See #72957 for details, but at present we get quite far into ILM execution before we confirm that the snapshot repository actually exists for the `searchable_snapshot` action, and then by the time we find out it doesn't, the only way out is to make it exist (which is unfortunate in the case of typos!).

A big part of the reason that ends up being such a pain is that we cache the `snapshot_repository` on the ILM custom metadata very early (at `GenerateSnapshotNameStep`) and then never change it, at the same time we also have asserts here that prevent that same step from running again.

My approach to solve this is two-fold: 

1. I've added validation to the `GenerateSnapshotNameStep` that checks to make sure that the snapshot repository exists before the step is allowed to execute. In the event that the repository doesn't exist, you can either fix the policy to reference a repository that does exist (the crucial typo case), or you can create the repository.

2. I've made it possible to execute `GenerateSnapshotNameStep` more than once. It still will not reset the generated snapshot name, but the rest of the logic executes regardless (and includes resetting the cached `snapshot_repository`). This allows for a manual escape hatch for clusters that are already in this error state and then upgrade to a version where this fix exists, namely they can `_ilm/move` to `generate-snapshot-name` in order to escape the error loop they'd otherwise find themselves in at `cleanup-snapshot`.

Note: @dakrone and I talked through a few approaches on this one, and... this PR is a totally different way that we didn't discuss at all. 😄

### CODE REVIEW

**Comment by joegallo** [comment]
I don't think the `elasticsearch-ci/bwc` is related to anything here, I'm assuming it'll sort itself out by merging in a recent-enough version of `master`.

**Review by martijnvg — APPROVED** [review]
This approach does look good to me.
I'm also curious in @dakrone's opinion.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/GenerateSnapshotNameStep.java [hunk 1]
```diff
@@ -61,29 +62,39 @@ public ClusterState performAction(Index index, ClusterState clusterState) {
...
-        if (validationException != null) {
-            logger.warn("unable to generate a snapshot name as part of policy [{}] for index [{}] due to [{}]",
-                policy, index.getName(), validationException.getMessage());
-            throw validationException;
+        LifecycleExecutionState lifecycleState = fromIndexMetadata(indexMetaData);
+
+        // validate that the snapshot repository exists -- because policies are refreshed on later retries, and because
+        // this fails prior to the snapshot repository being recorded in the ilm metadata, the policy can just be corrected
+        // and everything will pass on the subsequent retry
+        if (clusterState.metadata().custom(RepositoriesMetadata.TYPE, RepositoriesMetadata.EMPTY).repository(snapshotRepository) == null) {
```
[the remark this answers is not in the document]
Comment (joegallo) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/GenerateSnapshotNameStep.java:71]: +1 to that. @dakrone and I talked about adding that as a breaking change (so 8.0.0 only) at the rest layer. The scenario we're imagining is some sort of configuration as code that's going to put some templates and a policy and a snapshot repo as part of setting up ES, but where the policy put happens to be before the snapshot repo put -- currently that works, after our change it wouldn't.

**Review by dakrone — APPROVED** [review]
So this does LGTM, as it fixes the problem from an outside perspective. It's still possible to get wedged if you have an index in a step *after* the name generation that is using the wrong snapshot repo, but that should be an edge case once we have this validation.

As Martijn mentioned, +1 to adding the validation for existence to 8.0+ in a separate PR

**Comment by joegallo** [comment]
#78468 is up as a draft on validating this at policy creation/update time, too.

**Comment by joegallo** [comment]
@elasticmachine update branch

**Comment by joegallo** [comment]
@elasticmachine update branch

**Pull request merged**